### PR TITLE
Some refactor to make the code cleaner

### PR DESCRIPTION
**Pull request by ecastro98**
It would also be able to be used with composer, the package manager for php.

### CODE REVIEW

**Comment by qaisjp** [comment]
I've deleted commit 0205974 & fixed some other bugs to make the commit history cleaner. You will need to `git fetch` and then `git reset origin/feature/refactor --hard` so that you get the updated repo.

File: src/Model/Resources.php
Comment (shyim) [inline src/Model/Resources.php]: Maybe void?

File: composer.json
Comment (shyim) [inline composer.json]: Maybe use a client abstraction like http://httplug.io/ ?

File: src/Model/Resource.php
Comment (shyim) [inline src/Model/Resource.php]: Maybe make $server in constructor instead required?

File: src/Model/Resource.php
Comment (shyim) [inline src/Model/Resource.php]: Typehintings are missing :)

File: composer.json
Comment (shyim) [inline composer.json]: Autoload Namespace is registered two times on dev mode :)

File: src/Model/Resource.php
Comment (ecastro98) [reply src/Model/Resource.php]: This is because getInput, static method of Mta class, use the constructor. That's why I made it optional

File: composer.json
Comment (shyim) [inline composer.json]: This should be increased to 7.1.
You have used https://github.com/multitheftauto/mtasa-php-sdk/pull/2/files#diff-74763f9c84d84cc78a22ed38d80d45f9R25 nullable return types, which are introduced with PHP 7.1 :)

Comment (ecastro98) [reply composer.json]: Yeah, I was trying to make it compatible with PHP 7, didn't know about that. I will change it, thanks

Comment (Sasu98) [reply composer.json]: Or is it better to replace with false boolean? I mean, there is an important number of people using PHP 7. What do you think?

Comment (shyim) [reply composer.json]: I would increase just the php version to 7.1 and done.
PHP 7.0 security support already left out 😅

**Comment by ecastro98** [comment]
For some reason, WIP check dissapeared. Do not merge while 'WIP' is in title.

**Comment by ecastro98** [comment]
I think that is all for now. Thanks @shyim for your feedback. Can you please review this @qaisjp and @Deltanic ?

**Comment by qaisjp** [comment]
So much stuff! Awesome work. Heads up: it **will** take us a long time to merge this pull request.

File: examples/ajax/ajax_interface.php
Comment (qaisjp) [inline examples/ajax/ajax_interface.php]: heads up

**Comment by ecastro98** [comment]
More and more changes >:D @qaisjp @Deltanic

**Review by qaisjp — APPROVED** [review]
I gave it a read, this looks great!

Whilst I cannot judge whether or not the code is idiomatic (I am not a PHP developer), I couldn't find any obvious issues 👍 

My only comment would be that I am not sure if supporting the following syntax is a good idea:

```php
$response = $mta->someResource->callableFunction($arg1, $arg2, $arg3, ...);
```

My concern is that, if we add more ways to interact with a server, the possibility of clashing with existing resources might prevent us from adding more methods.

What do you think? It is totally your call whether or not to keep the functionality.

**Comment by ecastro98** [comment]
We have MtaService class where it happens the real interaction with the server. The magic method (in Resource class) will execute the method `call` from the service so if we add more ways to interact with a server, it should happen here with no problems.

Plus, it also adds some DX.

Although, I will remove the magic getter from Mta class and keep `$mta->getResource('someResource')` to get a resource because it restrict us to create public properties that could be important in the future.

So the final ways to call a resource function will be:
```php
$response = $mta->getResource('someResource')->call('callableFunction', $arg1, $arg2, $arg3, ...);
// or
$response = $mta->getResource('someResource')->callableFunction($arg1, $arg2, $arg3, ...);
```

**Comment by qaisjp** [comment]
> Although, I will remove the magic getter from Mta class and keep `$mta->getResource('someResource')` to get a resource because it restrict us to create public properties that could be important in the future.

This is actually what I was referring to. I didn't even notice the `$res->callableFunction(...)` shorthand!

I notice `Resource.php` has this:

```php
public function getName()
{
    return $this->name;
}
```

If there's a callable function `getName`, wouldn't you be required to use `$resource->call('getName', ...)` instead?

I agree it's nice to have the shorthand, but imo there's the same issue here. Maybe a nice way around that would be to support `$res->call->callableFunction(...)` or `$res->rpc->callableFunction(...)` instead?

**Comment by ecastro98** [comment]
Agree. I will add it tomorrow.

**Comment by qaisjp** [comment]
Oh, and one more thing, the license.

It looks like all of this is original work, so the license should be completely up to you.

I'd recommend either:
- [GNU AGPL v3](https://github.com/multitheftauto/mtasa-php-sdk/community/license/new?branch=master&template=agpl-3.0) ([choosealicense.com](https://choosealicense.com/licenses/agpl-3.0/)), or
- [MIT](https://github.com/multitheftauto/mtasa-php-sdk/community/license/new?branch=master&template=mit) ([choosealicense.com](https://choosealicense.com/licenses/mit/))

**Comment by ecastro98** [comment]
@qaisjp What about now?

**Review by qaisjp — APPROVED** [review]
Go for merge!

**Pull request merged**
